### DISCUSSION TITLE
How to read/decode UART data ?

### DISCUSSION THREAD

**Discussion by saepien** [micropython/micropython#10779] [ESP32] [answered]
Need help reading serialport data on my ESP32.

I'm following some example code I found and have this in my main.py:

`from machine import UART
import ubinascii

uart = UART(2, tx=17, rx=16)
uart.init(57600, bits=8, parity=None, stop=1)

while True:
    data = uart.readline()
    msg = ubinascii.unhexlify(data).decode('utf-8')

    if msg is not None:
        print(msg)`

But all I get is an error:
>>>
MPY: soft reboot
Traceback (most recent call last):
  File "main.py", line 13, in <module>
TypeError: object with buffer protocol required
MicroPython v1.19.1 on 2022-06-18; ESP32 module with ESP32
Type "help()" for more information.
>>>

When I omit the 'msg=' line and just print 'data' I get this:

b'\xe0\x00\x00\xe0\x00\x00\x00\xe0\x00\xe0\x00\x00\x00\x00\x00\x00\xe0\x00\xe0\x00\x00\xe0\x00\x00\x00\xe0\x00\x00\x00\x00\xe0\xe0\xe0\x00\x00\xe0\x00\x00\x00\x00\xe0\x00\x00\x00\xe0\x00\x00\xe0\x00\x00\xe0\x00\x00\xe0\x00\x00\xe0\x00\xe0\x00\x00\x00\xe0\x00\xe0\xe0\x00\xe0\xe0\xe0\x00\x00\xe0\x00\x00\x00\x00\xe0\x00\xe0\x00\x00\xe0\x00\x00\x00\xe0\x00\xe0\x00\x00\x00\xe0\x00\x00\xe0\x00\x00\x00\x00\xe0\x00\xe0\x00\x00\x00\xe0\xe0\x00\x00\x00\xe0\x00\x00\xe0\xe0\x00\x00\xe0\x00'

**Comment by davefes** [comment]
Drop UART in search.  Think there was a recent comment about readline versus read

**Reply by saepien** [reply]
Have tried both. Neither works. Can't seem to read any data from the serial device while I can easily and normally do so when the serial device is connected to my computer.

**Reply by davefes** [reply]
Ah, try:
```py
ser = UART(1, baudrate=9600, bits=8, parity=None, stop=1, rx=18, tx=19, timeout=300)
```
all in one line.  And with your parameters.

On re-reading that probably won't sort your problem.

**Reply by saepien** [reply]
Turns out the read buffer was too small. Go figure!?

**Reply by robert-hh** [reply]
> Can't seem to read any data from the serial device while I can easily and normally do so when the serial device is connected to my computer.

Is a GND wire connected between the two devices?

**Reply by saepien** [reply]
Yes, ground wasn't the issue. It was buffer size.

**Comment by saepien** [comment, marked as answer]
So basically all good now :)
Aside from another issue! Will post to new thread
